[VecOps] RVec 2.0: small buffer optimization based on LLVM SmallVector

eguiraud opened the pull request:
- [x] add ARCHITECTURE.md
- [x] use `fCapacity == -1` to indicate memory-adoption mode
- [x] switch `assert`s to `throw`s
- [x] expose the small buffer size as a template parameter (defaulted to `sizeof(T)*8 > 1024 ? 0 : 8` or similar, see also https://lists.llvm.org/pipermail/llvm-dev/2020-November/146613.html and the way they currently do it in LLVM: https://llvm.org/doxygen/SmallVector_8h_source.html#l01101)
- [x] re-check before/after benchmark runtimes (first measurements at https://eguiraud.web.cern.ch/eguiraud/decks/20201112_rvec_redesign_ppp )
- [x] unit test for exceptions thrown during construction or resizing (and add note about lack of exception safety in docs)
- [x] confirm that crediting of LLVM is ok (currently only in `math/vecops/ARCHITECTURE.md`)

eguiraud commented:
This is a reproducer for the Python failures:

```python
import ROOT
rvec = ROOT.VecOps.RVec("float")()
```

Stacktrace at the point of crash is:

```
#0  0x00007ffff726c389 in __gnu_cxx::new_allocator<CPyCppyy::PyCallable*>::construct<CPyCppyy::PyCallable*, CPyCppyy::PyCallable* const&> (this=0x7ffff72c93df, __p=0x5f3a3a726f746365) at /usr/include/c++/10.2.0/ext/new_allocator.h:150
150		{ ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
>>> bt
#0  0x00007ffff726c389 in __gnu_cxx::new_allocator<CPyCppyy::PyCallable*>::construct<CPyCppyy::PyCallable*, CPyCppyy::PyCallable* const&> (this=0x7ffff72c93df, __p=0x5f3a3a726f746365) at /usr/include/c++/10.2.0/ext/new_allocator.h:150
#1  0x00007ffff726bf06 in std::allocator_traits<std::allocator<CPyCppyy::PyCallable*> >::construct<CPyCppyy::PyCallable*, CPyCppyy::PyCallable* const&> (__a=..., __p=0x5f3a3a726f746365) at /usr/include/c++/10.2.0/bits/alloc_traits.h:512
#2  0x00007ffff726bc7a in std::vector<CPyCppyy::PyCallable*, std::allocator<CPyCppyy::PyCallable*> >::push_back (this=0x7ffff72c93df, __x=@0x7fffffffd630: 0x555559389d80) at /usr/include/c++/10.2.0/bits/stl_vector.h:1192
#3  0x00007ffff72760a9 in CPyCppyy::CPPOverload::AdoptMethod (this=0x7ffff73136e0 <CPyCppyy::CPPOverload_Type>, pc=0x555559389d80) at ../../../../../rvec2/bindings/pyroot/cppyy/CPyCppyy/src/CPPOverload.cxx:945
#4  0x00007ffff72b3ba6 in CPyCppyy::TemplateProxy::AdoptMethod (this=0x7fffe9222280, pc=0x555559389d80) at ../../../../../rvec2/bindings/pyroot/cppyy/CPyCppyy/src/TemplateProxy.cxx:111
#5  0x00007ffff72a4233 in CPyCppyy::BuildScopeProxyDict (scope=16, pyclass=0x555559379ca0) at ../../../../../rvec2/bindings/pyroot/cppyy/CPyCppyy/src/ProxyWrappers.cxx:254
#6  0x00007ffff72a6273 in CPyCppyy::CreateScopeProxy (name="ROOT::VecOps::RVec<float>", parent=0x555558e398c0) at ../../../../../rvec2/bindings/pyroot/cppyy/CPyCppyy/src/ProxyWrappers.cxx:676
#7  0x00007ffff72830eb in (anonymous namespace)::MakeCppTemplateClass (args=0x7ffff73ef040) at ../../../../../rvec2/bindings/pyroot/cppyy/CPyCppyy/src/CPyCppyyModule.cxx:368
#8  0x00007ffff7d33a92 in ?? () from /usr/lib/libpython3.9.so.1.0
#9  0x00007ffff7d3298f in PyObject_Call () from /usr/lib/libpython3.9.so.1.0
#10 0x00007ffff7d163da in _PyEval_EvalFrameDefault () from /usr/lib/libpython3.9.so.1.0
#11 0x00007ffff7d0f9ad in ?? () from /usr/lib/libpython3.9.so.1.0
#12 0x00007ffff7d21b0e in _PyFunction_Vectorcall () from /usr/lib/libpython3.9.so.1.0
#13 0x00007ffff7d19735 in _PyObject_FastCallDictTstate () from /usr/lib/libpython3.9.so.1.0
#14 0x00007ffff7d2eaa9 in _PyObject_Call_Prepend () from /usr/lib/libpython3.9.so.1.0
#15 0x00007ffff7e0de7a in ?? () from /usr/lib/libpython3.9.so.1.0
#16 0x00007ffff7d19e6d in _PyObject_MakeTpCall () from /usr/lib/libpython3.9.so.1.0
#17 0x00007ffff7d15b3a in _PyEval_EvalFrameDefault () from /usr/lib/libpython3.9.so.1.0
#18 0x00007ffff7d0f9ad in ?? () from /usr/lib/libpython3.9.so.1.0
#19 0x00007ffff7d0f371 in _PyEval_EvalCodeWithName () from /usr/lib/libpython3.9.so.1.0
#20 0x00007ffff7dd1f83 in PyEval_EvalCode () from /usr/lib/libpython3.9.so.1.0
#21 0x00007ffff7de23dd in ?? () from /usr/lib/libpython3.9.so.1.0
#22 0x00007ffff7dddc7b in ?? () from /usr/lib/libpython3.9.so.1.0
#23 0x00007ffff7c7f20b in ?? () from /usr/lib/libpython3.9.so.1.0
#24 0x00007ffff7c7e5e3 in PyRun_SimpleFileExFlags () from /usr/lib/libpython3.9.so.1.0
#25 0x00007ffff7df3dba in Py_RunMain () from /usr/lib/libpython3.9.so.1.0
#26 0x00007ffff7dc4fa9 in Py_BytesMain () from /usr/lib/libpython3.9.so.1.0
#27 0x00007ffff7a41b25 in __libc_start_main () from /usr/lib/libc.so.6
#28 0x000055555555504e in _start ()
```

`gdb` shows some problems in `CPyCppyy::CPPOverload::fMethodInfo`:

```
#3  0x00007ffff72760a9 in CPyCppyy::CPPOverload::AdoptMethod (this=0x7ffff73136e0 <CPyCppyy::CPPOverload_Type>, pc=0x555559389d80) at ../../../../../rvec2/bindings/pyroot/cppyy/CPyCppyy/src/CPPOverload.cxx:945
945	    fMethodInfo->fMethods.push_back(pc);
>>> p *fMethodInfo
```

and `valgrind` agrees:

```
==216976== Invalid write of size 8
==216976==    at 0x5DC0389: void __gnu_cxx::new_allocator<CPyCppyy::PyCallable*>::construct<CPyCppyy::PyCallable*, CPyCppyy::PyCallable* const&>(CPyCppyy::PyCallable**, CPyCppyy::PyCallable* const&) (new_allocator.h:150)
==216976==    by 0x5DBFF05: void std::allocator_traits<std::allocator<CPyCppyy::PyCallable*> >::construct<CPyCppyy::PyCallable*, CPyCppyy::PyCallable* const&>(std::allocator<CPyCppyy::PyCallable*>&, CPyCppyy::PyCallable**, CPyCppyy::PyCallable* const&) (alloc_traits.h:512)
==216976==    by 0x5DBFC79: std::vector<CPyCppyy::PyCallable*, std::allocator<CPyCppyy::PyCallable*> >::push_back(CPyCppyy::PyCallable* const&) (stl_vector.h:1192)
==216976==    by 0x5DCA0A8: CPyCppyy::CPPOverload::AdoptMethod(CPyCppyy::PyCallable*) (CPPOverload.cxx:945)
==216976==    by 0x5E07BA5: CPyCppyy::TemplateProxy::AdoptMethod(CPyCppyy::PyCallable*) (TemplateProxy.cxx:111)
==216976==    by 0x5DF8232: CPyCppyy::BuildScopeProxyDict(unsigned long, _object*) (ProxyWrappers.cxx:254)
==216976==    by 0x5DFA272: CPyCppyy::CreateScopeProxy(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, _object*) (ProxyWrappers.cxx:676)
==216976==    by 0x5DD70EA: (anonymous namespace)::MakeCppTemplateClass(_object*, _object*) (CPyCppyyModule.cxx:368)
==216976==    by 0x49B1A91: ??? (in /usr/lib/libpython3.9.so.1.0)
==216976==    by 0x49B098E: PyObject_Call (in /usr/lib/libpython3.9.so.1.0)
==216976==    by 0x49943D9: _PyEval_EvalFrameDefault (in /usr/lib/libpython3.9.so.1.0)
==216976==    by 0x498D9AC: ??? (in /usr/lib/libpython3.9.so.1.0)
==216976==  Address 0x5f3a3a726f746365 is not stack'd, malloc'd or (recently) free'd
```

stwunsch commented:
I've built with clang10 and debug info and rebased on top of master. Then I get following additional info:

```bash
[wunsch@portal1 build_native]$ python -c 'import ROOT; rvec = ROOT.VecOps.RVec("float")()'
input_line_14:1:64: error: redefinition of 'is_equal'
namespace __cppyy_internal { template<class C1, class C2> bool is_equal(const C1& c1, const C2& c2) { return (bool)(c1 == c2); } }
                                                               ^
input_line_10:1:64: note: previous definition is here
namespace __cppyy_internal { template<class C1, class C2> bool is_equal(const C1& c1, const C2& c2) { return (bool)(c1 == c2); } }
                                                               ^
input_line_15:1:64: error: redefinition of 'is_not_equal'
namespace __cppyy_internal { template<class C1, class C2> bool is_not_equal(const C1& c1, const C2& c2) { return (bool)(c1 != c2); } }
                                                               ^
input_line_11:1:64: note: previous definition is here
namespace __cppyy_internal { template<class C1, class C2> bool is_not_equal(const C1& c1, const C2& c2) { return (bool)(c1 != c2); } }
                                                               ^
 *** Break *** segmentation violation



===========================================================
There was a crash (kSigSegmentationViolation).
This is the entire stack trace of all threads:
===========================================================
#0  0x00007f0ac351e46c in waitpid () from /lib64/libc.so.6
#1  0x00007f0ac349bf62 in do_system () from /lib64/libc.so.6
#2  0x00007f0abbefaf19 in TUnixSystem::Exec (this=0x193ebd0, shellcmd=0x5886a50 "/work/wunsch/root_rvec/build_native/etc/gdb-backtrace.sh 3925717 1>&2") at /work/wunsch/root_rvec/core/unix/src/TUnixSystem.cxx:2120
#3  0x00007f0abbefba11 in TUnixSystem::StackTrace (this=0x193ebd0) at /work/wunsch/root_rvec/core/unix/src/TUnixSystem.cxx:2411
#4  0x00007f0aa6eb2e91 in (anonymous namespace)::TExceptionHandlerImp::HandleException(int) () from /cvmfs/sft-nightlies.cern.ch/lcg/nightlies/dev3/Mon/ROOT/HEAD/x86_64-centos7-clang10-opt/lib/libcppyy_backend3_8.so
#5  0x00007f0abbf01176 in TUnixSystem::DispatchSignals (this=0x193ebd0, sig=kSigSegmentationViolation) at /work/wunsch/root_rvec/core/unix/src/TUnixSystem.cxx:3644
#6  0x00007f0abbef5f81 in SigHandler (sig=kSigSegmentationViolation) at /work/wunsch/root_rvec/core/unix/src/TUnixSystem.cxx:407
#7  0x00007f0abbf01364 in sighandler (sig=11) at /work/wunsch/root_rvec/core/unix/src/TUnixSystem.cxx:3620
#8  <signal handler called>
#9  0x00007f0abc71ec1c in __gnu_cxx::new_allocator<CPyCppyy::PyCallable*>::construct<CPyCppyy::PyCallable*, CPyCppyy::PyCallable* const&> (this=0x7f0abc784120, __p=0x5f746f6f722f6863, __args=
0x7ffeaf819970: 0x58665b0) at /cvmfs/sft.cern.ch/lcg/releases/gcc/9.2.0-afc57/x86_64-centos7/lib/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/ext/new_allocator.h:147
#10 0x00007f0abc71ea1d in std::allocator_traits<std::allocator<CPyCppyy::PyCallable*> >::construct<CPyCppyy::PyCallable*, CPyCppyy::PyCallable* const&> (__a=..., __p=0x5f746f6f722f6863, __args=
0x7ffeaf819970: 0x58665b0) at /cvmfs/sft.cern.ch/lcg/releases/gcc/9.2.0-afc57/x86_64-centos7/lib/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/alloc_traits.h:484
#11 0x00007f0abc71e81e in std::vector<CPyCppyy::PyCallable*, std::allocator<CPyCppyy::PyCallable*> >::push_back (this=0x7f0abc784120, __x=
0x7ffeaf819970: 0x58665b0) at /cvmfs/sft.cern.ch/lcg/releases/gcc/9.2.0-afc57/x86_64-centos7/lib/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/stl_vector.h:1189
#12 0x00007f0abc72709f in CPyCppyy::CPPOverload::AdoptMethod (this=0x7f0abc9caed8 <CPyCppyy::CPPOverload_Type>, pc=0x58665b0) at /work/wunsch/root_rvec/bindings/pyroot/cppyy/CPyCppyy/src/CPPOverload.cxx:945
#13 0x00007f0abc76e7f0 in CPyCppyy::TemplateProxy::AdoptMethod (this=0x7f0aaca317c0, pc=0x58665b0) at /work/wunsch/root_rvec/bindings/pyroot/cppyy/CPyCppyy/src/TemplateProxy.cxx:111
#14 0x00007f0abc75ed8c in CPyCppyy::BuildScopeProxyDict (scope=14, pyclass=0x5859310) at /work/wunsch/root_rvec/bindings/pyroot/cppyy/CPyCppyy/src/ProxyWrappers.cxx:254
#15 0x00007f0abc75d3de in CPyCppyy::CreateScopeProxy (name=..., parent=0x57021d0) at /work/wunsch/root_rvec/bindings/pyroot/cppyy/CPyCppyy/src/ProxyWrappers.cxx:676
#16 0x00007f0abc7378d8 in (anonymous namespace)::MakeCppTemplateClass (args=0x7f0abca483c0) at /work/wunsch/root_rvec/bindings/pyroot/cppyy/CPyCppyy/src/CPyCppyyModule.cxx:368
#17 0x00007f0ac440cd1a in cfunction_call_varargs (func=0x7f0abca8aef0, args=0x7f0abca483c0, kwargs=<optimized out>) at /workspace/build/externals/Python-3.8.6/src/Python/3.8.6/Objects/call.c:757
#18 0x00007f0ac440d834 in PyCFunction_Call (func=0x7ffeaf819970, args=0x5f746f6f722f6863, kwargs=0x7ffeaf819970) at /workspace/build/externals/Python-3.8.6/src/Python/3.8.6/Objects/call.c:772
#19 0x00007f0ac44d7b0c in do_call_core (tstate=0x1884330, func=0x7f0abca8aef0, callargs=0x7f0abca483c0, kwdict=0x0) at /workspace/build/externals/Python-3.8.6/src/Python/3.8.6/Python/ceval.c:4983
```

eguiraud commented:
That looks like a different error rather than additional info :thinking:

stwunsch commented:
With this patch, you can find out which method causes the issue:

```diff
--- a/bindings/pyroot/cppyy/CPyCppyy/src/CPPOverload.cxx
+++ b/bindings/pyroot/cppyy/CPyCppyy/src/CPPOverload.cxx
@@ -20,6 +20,7 @@
 #include <algorithm>
 #include <sstream>
 #include <vector>
+#include <iostream>


 namespace CPyCppyy {
@@ -941,6 +942,13 @@ void CPyCppyy::CPPOverload::Set(const std::string& name, std::vector<PyCallable*
 //----------------------------------------------------------------------------
 void CPyCppyy::CPPOverload::AdoptMethod(PyCallable* pc)
 {
+    PyObject_Print(pc->GetPrototype(), stdout, 0);
+    std::cout << std::endl;
+
+    PyObject_Print(pc->GetSignature(), stdout, 0);
+    std::cout << std::endl;
+
+    std::cout << std::endl;
 // Fill in the data of a freshly created method proxy.
     fMethodInfo->fMethods.push_back(pc);
     fMethodInfo->fFlags &= ~CallContext::kIsSorted;
```

It gets you to the point that you have multiple method with the same signature:

```bash
...

'float& ROOT::VecOps::RVec<float>::operator[](unsigned long idx)'
'(unsigned long idx)'

'float& ROOT::VecOps::RVec<float>::operator[](unsigned long idx)'
'(unsigned long idx)'
```

eguiraud commented:
I don't understand, the problem is not the "redefinition of `is_equal`" ?

stwunsch commented:
My best guess is that cppyy sees a method two times and then defines the equal operator in the jit twice. The iteration over the methods happens here (the counter is there to set a breakpoint more easily and continue until the actual critical part happens):

```diff
diff --git a/bindings/pyroot/cppyy/CPyCppyy/src/ProxyWrappers.cxx b/bindings/pyroot/cppyy/CPyCppyy/src/ProxyWrappers.cxx
index 23e7ac8..70b662c 100644
--- a/bindings/pyroot/cppyy/CPyCppyy/src/ProxyWrappers.cxx
+++ b/bindings/pyroot/cppyy/CPyCppyy/src/ProxyWrappers.cxx
@@ -26,6 +26,7 @@
 #include <set>
 #include <string>
 #include <vector>
+#include <iostream>


 //- data _______________________________________________________________________
@@ -141,6 +142,8 @@ static inline void sync_templates(
     Py_DECREF(pyname);
 }

+static auto counter = 0u;
+
 static int BuildScopeProxyDict(Cppyy::TCppScope_t scope, PyObject* pyclass)
 {
 // Collect methods and data for the given scope, and add them to the given python
@@ -169,6 +172,9 @@ static int BuildScopeProxyDict(Cppyy::TCppScope_t scope, PyObject* pyclass)

     // process the method based on its name
         std::string mtCppName = Cppyy::GetMethodName(method);
+        std::cout << "Counter: " << counter << std::endl;
+        counter++;
+        std::cout << "Method name: " << mtCppName << std::endl;

     // special case trackers
         bool setupSetItem = false;
 ```

stwunsch commented:
More info: The `is_equal` is jitted on the fly in the method here:
https://github.com/root-project/root/blob/master/bindings/pyroot/cppyy/CPyCppyy/src/Utility.cxx#L301

And this is again called in various places in https://github.com/root-project/root/blob/master/bindings/pyroot/cppyy/CPyCppyy/src/CPPInstance.cxx

And a `CPPInstance` is one of the core classes :)

stwunsch commented on math/vecops/inc/ROOT/RVec.hxx:
```diff
@@ -271,160 +259,1076 @@ hpt->Draw();
+      return *this;
+   }
+
+   using SmallVectorTemplateCommon<T>::operator[];
```
This using declaration may be the culprit? We have a rather complicated past in terms of using declarations and jitting ;)

stwunsch commented:
The `is_equal` issue may be a red herring :/

It's some weirdness which includes inheritance with also a templated method in the derived class.

Here is the most minimal reproducer I could find:

```python
import cppyy

cppyy.cppdef('''
template <typename T>
struct Base {
    T data;
    T& operator[](int idx) { return data; }
};

template <typename T>
struct Vec : public Base<T> {
    T data;

    // we have two options to trigger the issue:

    // 1. Including this line makes it break
    //using Base<T>::operator[];

    // 2. Or including this line
    // Note that you have to derive Vec from Base to make it break!
    T& operator[](int idx) { return data; }

    // Note that you have to have the template here!
    template <typename V>
    T& operator[](T idx) {
        return data;
    }
};
''')

vec = cppyy.gbl.Vec['float']()
print(vec, vec[0])
vec[0] = 42.0
print(vec, vec[0])
```

And here the stacktrace, which is the same as above:

```bash
#9  0x00007f09d1f0bc1c in __gnu_cxx::new_allocator<CPyCppyy::PyCallable*>::construct<CPyCppyy::PyCallable*, CPyCppyy::PyCallable* const&> (this=0x7f09d1f71120, __p=0x5f746f6f722f6863, __args=
0x7ffd74585880: 0x5f009c0) at /cvmfs/sft.cern.ch/lcg/releases/gcc/9.2.0-afc57/x86_64-centos7/lib/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/ext/new_allocator.h:147
#10 0x00007f09d1f0ba1d in std::allocator_traits<std::allocator<CPyCppyy::PyCallable*> >::construct<CPyCppyy::PyCallable*, CPyCppyy::PyCallable* const&> (__a=..., __p=0x5f746f6f722f6863, __args=
0x7ffd74585880: 0x5f009c0) at /cvmfs/sft.cern.ch/lcg/releases/gcc/9.2.0-afc57/x86_64-centos7/lib/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/alloc_traits.h:484
#11 0x00007f09d1f0b81e in std::vector<CPyCppyy::PyCallable*, std::allocator<CPyCppyy::PyCallable*> >::push_back (this=0x7f09d1f71120, __x=
0x7ffd74585880: 0x5f009c0) at /cvmfs/sft.cern.ch/lcg/releases/gcc/9.2.0-afc57/x86_64-centos7/lib/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/stl_vector.h:1189
#12 0x00007f09d1f1409f in CPyCppyy::CPPOverload::AdoptMethod (this=0x7f09d21b7ed8 <CPyCppyy::CPPOverload_Type>, pc=0x5f009c0) at /work/wunsch/root_rvec/bindings/pyroot/cppyy/CPyCppyy/src/CPPOverload.cxx:945
#13 0x00007f09d1f5b7f0 in CPyCppyy::TemplateProxy::AdoptMethod (this=0x7f09c568c490, pc=0x5f009c0) at /work/wunsch/root_rvec/bindings/pyroot/cppyy/CPyCppyy/src/TemplateProxy.cxx:111
#14 0x00007f09d1f4bd8c in CPyCppyy::BuildScopeProxyDict (scope=8, pyclass=0x5f31f60) at /work/wunsch/root_rvec/bindings/pyroot/cppyy/CPyCppyy/src/ProxyWrappers.cxx:254
#15 0x00007f09d1f4a3de in CPyCppyy::CreateScopeProxy (name=..., parent=0x7f09d223a590) at /work/wunsch/root_rvec/bindings/pyroot/cppyy/CPyCppyy/src/ProxyWrappers.cxx:676
#16 0x00007f09d1f248d8 in (anonymous namespace)::MakeCppTemplateClass (args=0x7f09d2237380) at /work/wunsch/root_rvec/bindings/pyroot/cppyy/CPyCppyy/src/CPyCppyyModule.cxx:368
#17 0x00007f09d9bf9d1a in cfunction_call_varargs (func=0x7f09d21fe360, args=0x7f09d2237380, kwargs=<optimized out>) at /workspace/build/externals/Python-3.8.6/src/Python/3.8.6/Objects/call.c:757
#18 0x00007f09d9bfa834 in PyCFunction_Call (func=0x7ffd74585880, args=0x5f746f6f722f6863, kwargs=0x7ffd74585880) at /workspace/build/externals/Python-3.8.6/src/Python/3.8.6/Objects/call.c:772
#19 0x00007f09d9cc4b0c in do_call_core (tstate=0x23ba020, func=0x7f09d21fe360, callargs=0x7f09d2237380, kwdict=0x0) at /workspace/build/externals/Python-3.8.6/src/Python/3.8.6/Python/ceval.c:4983
#20 _PyEval_EvalFrameDefault (f=0x7f09da1609f0, throwflag=<optimized out>) at /workspace/build/externals/Python-3.8.6/src/Python/3.8.6/Python/ceval.c:3559
#21 0x00007f09d9cc8863 in PyEval_EvalFrameEx (f=0x7f09da1609f0, throwflag=0) at /workspace/build/externals/Python-3.8.6/src/Python/3.8.6/Python/ceval.c:741
```

stwunsch commented:
There we go, the best reproducer so far:

```python
import cppyy

cppyy.cppdef('''
struct Base {
    float data;
    float& operator[](int idx) { return data; }
};

struct Derived: public Base {
    float data;

    // Note that you have to derive from Base to make it break!
    float& operator[](int idx) { return data; }

    // Note that you must put the template here!
    template <typename T>
    float& operator[](int idx) { return data; }
};
''')

cppyy.gbl.Derived()
```

stwunsch commented:
There we go, I've made an issue upstream: https://bitbucket.org/wlav/cppyy/issues/334/segfault-when-creating-proxy-to-derived

stwunsch commented:
@eguiraud In my built, this patch is a workaround for the segfault discussed above:

```patch
diff --git a/math/vecops/inc/ROOT/RVec.hxx b/math/vecops/inc/ROOT/RVec.hxx
index d9adca8..5900743 100644
--- a/math/vecops/inc/ROOT/RVec.hxx
+++ b/math/vecops/inc/ROOT/RVec.hxx
@@ -414,17 +414,6 @@ public:
    /// Return a pointer to the vector's buffer, even if empty().
    const_pointer data() const { return const_pointer(begin()); }

-   reference operator[](size_type idx)
-   {
-      assert(idx < size());
-      return begin()[idx];
-   }
-   const_reference operator[](size_type idx) const
-   {
-      assert(idx < size());
-      return begin()[idx];
-   }
-
    reference front()
    {
       assert(!empty());
@@ -1266,7 +1255,21 @@ public:
       return *this;
    }

-   using SmallVectorTemplateCommon<T>::operator[];
+   using reference = typename SmallVectorTemplateCommon<T>::reference;
+   using const_reference = typename SmallVectorTemplateCommon<T>::const_reference;
+   using size_type = typename SmallVectorTemplateCommon<T>::size_type;
+   using SmallVectorTemplateCommon<T>::begin;
+
+   reference operator[](size_type idx)
+   {
+      assert(idx < size());
+      return begin()[idx];
+   }
+   const_reference operator[](size_type idx) const
+   {
+      assert(idx < size());
+      return begin()[idx];
+   }

    template <typename V, typename = std::enable_if<std::is_convertible<V, bool>::value>>
    RVec operator[](const RVec<V> &conds) const
    ```

stwunsch commented:
Hm, but there is now a segfault in `tutorial-vecops-vo004_SortAndSelect-py` :/

wlav commented:
Aside, while discussing #7201 and constructing an example, I ran into that same `is_equal` redefinition mentioned above. All you need to do is this (from a ROOT build, not master):
```
>>> import cppyy
>>> cppyy.gbl.TPython.Eval("1+1")
```
and you get that error, which in turn tells me it's a pure ROOT issue, not a cppyy one, which doesn't have `TPython`. (You're certainly off the mark above in thinking it's "jitted on the fly" in Utilitiy.cxx, even in the ROOT fork of cppyy: what you point to is only the instantiation, not the declaration.)

stwunsch commented:
> Aside, while discussing #7201 and constructing an example, I ran into that same `is_equal` redefinition mentioned above. All you need to do is this (from a ROOT build, not master):
> 
> ```
> >>> import cppyy
> >>> cppyy.gbl.TPython.Eval("1+1")
> ```
> 
> and you get that error, which in turn tells me it's a pure ROOT issue, not a cppyy one, which doesn't have `TPython`. (You're certainly off the mark above in thinking it's "jitted on the fly" in Utilitiy.cxx, even in the ROOT fork of cppyy: what you point to is only the instantiation, not the declaration.)

Thanks for pointing that out! Indeed, the redefinition is a red herring and not connected to the issues with accessing the new `RVec` impl from Python. I push this discussion in another thread.

wlav commented:
Oh boy, this is the cause of your double declaration (run with `LD_DEBUG=files`):
```
$ LD_DEBUG=files python rbug2.py
[.. snip ..]
   3787500:         calling init: /home/wlav/pypy-dev/rbuild/lib/libcppyy_backend2_7.so
[.. snip ..]
   3787500:         calling init: /home/wlav/pypy-dev/rbuild/lib/libcppyy_backend3_8.so
[.. snip ..]
```
both of which will construct an `ApplicationStarter` object, running the declares through `ProcessLine()`.

What is even more bananas, is that cppyy_backend is _designed to be independent_ of Python and hence of Python version ... It's the entry point for `CPyCppyy`, `_cppyy` (in PyPy), and the D bindings into `libCling` and friends. As such, it relies on an ABI, not an API ...

It may be a red herring (since the original problem can be reproduced in master), but by duplicating a whole bunch of global objects, you'll be in a world of hurt no matter what other problems there are.

stwunsch commented:
Another thing that is broken with the new `RVec` is iteration:

```python
[wunsch@portal1 build_root]$ cat rep.py
import ROOT

rvec = ROOT.VecOps.RVec['float']((1, 2, 3))
c = 0
for x in rvec:
    print(x)
    c += 1
    if c > 5:
        break
```

```bash
1.0
2.0
3.0
3.518908852129243e-05
0.0
0.0
```

I vaguely remember this issue. We had something similar before (and that's why we have a test for it). I'll have a look!

stwunsch commented on math/vecops/inc/ROOT/RVec.hxx:
```diff
@@ -271,160 +259,1076 @@ hpt->Draw();
+   using size_type = size_t;
+   using difference_type = ptrdiff_t;
+   using value_type = T;
+   using iterator = T *;
```
The issue in the PyROOT iteration comes due to this. `RVec::begin()` does not return an iterator, it returns a `T *`!

@eguiraud There's a design decision behind this?

eguiraud commented:
Sorry too much info, too little bandwidth :sweat_smile: If I understand correctly:

- the `is_equal` redefinition is a real issue but unrelated to RVec 2.0 (thanks @wlav for pointing that out) and now tracked by #7678 
- you have a patch that works around the cppyy bug with `operator[]`
- iteration is broken because `RVec::begin` returns a raw pointer

In principle `begin` can absolutely return a raw pointer (I _think_ LLVM's SmallVector was designed like that for simplicity, to avoid having to define a custom iterator type). But raw pointers make for fine iterators :smile: Seems like a bug in PyROOT?

stwunsch commented:
> In principle `begin` can absolutely return a raw pointer (I _think_ LLVM's SmallVector was designed like that for simplicity, to avoid having to define a custom iterator type). But raw pointers make for fine iterators 😄 Seems like a bug in PyROOT?

I just had a quick look, but it seems to me that this is a corner case which doesn't trigger any mechanic for a proper pythonization, which enables the `for x in rvec` syntax in Python. Let's see, the final word is not yet spoken ;)

stwunsch commented:
Alright, it's again that PyROOT doesn't pick up methods from the inheritance hierarchy. Indeed, a raw pointer is totally find as an iterator. In this case, PyROOT doesn't pick up the`size` method and therefore just iterates forever. Here is a patch which makes it work:

```diff
diff --git a/math/vecops/inc/ROOT/RVec.hxx b/math/vecops/inc/ROOT/RVec.hxx
index 5900743..165b98a 100644
--- a/math/vecops/inc/ROOT/RVec.hxx
+++ b/math/vecops/inc/ROOT/RVec.hxx
@@ -1259,6 +1259,7 @@ public:
    using const_reference = typename SmallVectorTemplateCommon<T>::const_reference;
    using size_type = typename SmallVectorTemplateCommon<T>::size_type;
    using SmallVectorTemplateCommon<T>::begin;
+   using SmallVectorTemplateCommon<T>::size;

    reference operator[](size_type idx)
    {
    ```

stwunsch commented:
All tests pass for `ctest -R '(rvec|vecops|numpy|rdf|dataframe)'`, I guess that's fine for now?

eguiraud commented:
Absolutely, thank you very much Stefan! Please push your patches to this branch.

Do we have a separate issue to track this problem with PyROOT? Even just to document the issue, the workaround, the linked cppyy issue etc.

stwunsch commented:
I'll take care of pushing the fixes and the issue.

stwunsch commented:
> Build failed on ROOT-fedora31/noimt.
> Running on root-fedora-31-2.cern.ch:/home/sftnight/build/workspace/root-pullrequests-build
> [See console output](https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-build/112046/console).
> 
> ### Failing tests:
> * [projectroot.bindings.pyroot.pythonizations.test.pyunittests_pyroot_pyz_pretty_printing](https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-build/112046/testReport/projectroot.bindings.pyroot.pythonizations/test/pyunittests_pyroot_pyz_pretty_printing/)

This test fails because the pretty printing mechanism of cling is not picked up for the RVec 2.0. Actually I don't see why because the `operator<<` overload is there (see bottom of `RVec.hxx`).

stwunsch commented:
Uhm what also doesn't work is the following:

```python
>>> import ROOT
>>> x = ROOT.RVec['float']()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/work/wunsch/root_rvec/build_root/lib/ROOT/_facade.py", line 222, in _getattr
    return getattr(self, name)
  File "/work/wunsch/root_rvec/build_root/lib/ROOT/_facade.py", line 192, in _fallback_getattr
    raise AttributeError("Failed to get attribute {} from ROOT".format(name))
AttributeError: Failed to get attribute RVec from ROOT
```

But we don't have any test coverage because all tests were written before we made `RVec` accessible from the `ROOT` namespace. I'm going to fix that!

vgvassilev commented:
I think we should consider making the LLVMSupport library (which provides SmallVector & co) available in core and thus to the rest of ROOT.

cc: @Axel-Naumann

wlav commented:
> Edit: It's a long standing issue, see #7687 or #5361 or https://sft.its.cern.ch/jira/browse/ROOT-10688.

And all work fine in cppyy master (not that it is or ever was a cppyy issue).

eguiraud commented:
@vgvassilev that's fairly unrelated? It would not solve our cppyy-related issues and here we anyway patch `SmallVector` to add the ability to adopt an existing memory buffer, which is required for no-copy (or less-copy) I/O within RDataFrame.

stwunsch commented:
> > Edit: It's a long standing issue, see #7687 or #5361 or https://sft.its.cern.ch/jira/browse/ROOT-10688.
> 
> And all work fine in cppyy master (not that it is or ever was a cppyy issue).

This is discussed [here](https://sft.its.cern.ch/jira/browse/ROOT-10688).

vgvassilev commented:
> @vgvassilev that's fairly unrelated?

I am confused -- did we copy code from llvm? If we did I'd think that's quite related.

> It would not solve our cppyy-related issues and here we anyway patch `SmallVector` to add the ability to adopt an existing memory buffer, which is required for no-copy (or less-copy) I/O within RDataFrame.

I thought we have a way to do that with TMemFile already.

eguiraud commented:
@vgvassilev yes the bulk of the code is from LLVM SmallVector but see my comment above: we patch it to add a feature, and it does not work with (our version of) cppyy as-is. So exposing LLVM's SmallVectore as a core library would not help here, we'd still need to roll our own SmallVector-based RVec.

> I thought we have a way to do that with TMemFile already.

It's a different "level" of adoption: this is after deserialization, TMemFile might make an on-disk blob available as a memory buffer but that would be before deserialization.

eguiraud commented:
@phsft-bot build please

eguiraud commented:
Ah I broke pretty-printing of `std::map` :man_facepalming:

bellenot commented:
@eguiraud FYI the Windows node rebooted, so I killed the build to avoid a too long time-out

stwunsch commented:
> Ah I broke pretty-printing of `std::map` 🤦‍♂️

I'm proud of our test coverage :)

eguiraud commented:
Fixed pretty-printing, remaining issues seem related to autoloading:

```
IncrementalExecutor::executeFunction: symbol '_ZN4ROOT6VecOps15SmallVectorBaseIjE8grow_podEPvmm' unresolved while linking [cling interface function]!
You are probably missing the definition of ROOT::VecOps::SmallVectorBase<unsigned int>::grow_pod(void*, unsigned long, unsigned long)
Maybe you need to load the corresponding shared library?
```

Not sure what's going on on Windows, will have to investigate.

bellenot commented:
> Fixed pretty-printing, remaining issues seem related to autoloading:
> 
> ```
> IncrementalExecutor::executeFunction: symbol '_ZN4ROOT6VecOps15SmallVectorBaseIjE8grow_podEPvmm' unresolved while linking [cling interface function]!
> You are probably missing the definition of ROOT::VecOps::SmallVectorBase<unsigned int>::grow_pod(void*, unsigned long, unsigned long)
> Maybe you need to load the corresponding shared library?
> ```
> 
> Not sure what's going on on Windows, will have to investigate.

Could be due to an an inline function?

vgvassilev commented:
> @vgvassilev yes the bulk of the code is from LLVM SmallVector but see my comment above: we patch it to add a feature, and it does not work with (our version of) cppyy as-is. So exposing LLVM's SmallVectore as a core library would not help here, we'd still need to roll our own SmallVector-based RVec.
> 
> > I thought we have a way to do that with TMemFile already.
> 
> It's a different "level" of adoption: this is after deserialization, TMemFile might make an on-disk blob available as a memory buffer but that would be before deserialization.

Another thing to keep in mind is that the performance of the llvm::SmallVector comes from the fact it is not exception safe as llvm does not compile with exceptions. I suspect that may not apply for many parts of ROOT.

eguiraud commented:
> the performance of the llvm::SmallVector comes from the fact it is not exception safe

Can you elaborate a bit? For us, the performance boost in RVec 2.0 comes from the small buffer optimization, which reduces the amount of heap allocations during analysis event loops.

vgvassilev commented:
> > the performance of the llvm::SmallVector comes from the fact it is not exception safe
> 
> Can you elaborate a bit? For us, the performance boost in RVec 2.0 comes from the small buffer optimization, which reduces the amount of heap allocations during analysis event loops.

The implementation of SmallVector in llvm assumes it is used in codebases without exception support (like llvm itself). Thus the implementation can make some assumptions which the general std::vector cannot. See some terse documentation [here](https://llvm.org/doxygen/classllvm_1_1SmallVector.html).

There are two things I'd encourage you to try and see if you can reach the same level of small buffer optimization as you got now:
  * Make a wrapper class which always takes an assumed size over std::vector and call std::vector::reserve
  * Make a wrapper class which takes a template argument assumed size and try to allocate the assumed size on the stack, if it exceeds that, call underlying std::vector and move things there.

If either gets similar performance that'd be probably better approach in my opinion as it'd be exception safe.

eguiraud commented:
@vgvassilev if "not exception safe" here means w.r.t. `bad_allocs`, ROOT in general does not care (nothing works in that case anyway). If "not exception safe" means "might leak if constructors of elements throw", I don't think we care too much:

- RVec is mostly used with fundamental types  
- anyway that's not a situation in which analyses can/want to keep running, typically

Regarding your alternative proposals: we can't use a vanilla `std::vector` as a workhorse because we want to be able to adopt existing memory buffers. That would require std::vector + a custom allocator (like in the current implementation of RVec) which is fairly tricky to implement correctly (in fact, the current implementation has [broken usecases](https://github.com/root-project/root/issues/6796), and the more complex class layout makes it harder for RNTuple to perform I/O. `std::vector<bool>` weirdness complicates matters (and I/O) further. So the alternative is basically re-implementing vector-with-adoption-and-sbo ourselves.

All considered I think we could be ok with leaving the object in a "bad" state in case the construction of vector elements throws (which I think is what "not exception safe" means in this context), for the two reasons mentioned above -- but I'd like @Axel-Naumann to make the call.

Axel-Naumann commented:
Thanks, @vgvassilev

>  it is not exception safe

Instead of re-implementing a vector with small buffer optimization (along the lines you describe) we believe that adjusting this adapted small vector is a more viable / cheap solution. We also don't expect the lack of exception safety to be a major concern given the use cases of `RVec`. And if I am proven wrong and it becomes a concern then we're in a position to address that.

eguiraud commented:
@phsft-bot build just on windows10/cxx14

eguiraud commented:
@phsft-bot build just on windows10/cxx14

bellenot commented:
@phsft-bot  build just on windows10/cxx14

eguiraud commented:
@phsft-bot build just on windows10/cxx14

eguiraud commented:
This is finally green and ready for a first round of review.

Axel-Naumann commented on math/vecops/inc/LinkDef.h:
```diff
@@ -1,5 +1,5 @@
 /*************************************************************************
- * Copyright (C) 1995-2018, Rene Brun and Fons Rademakers.               *
+ * Copyright (C) 1995-2020, Rene Brun and Fons Rademakers.               *
```
That's a surprising year ;-)

Axel-Naumann commented on math/vecops/inc/ROOT/RVec.hxx:
```diff
@@ -276,163 +1214,179 @@ auto hpt = d.Define("pt", "sqrt(pxs * pxs + pys * pys)[E>200]")
 
-   template <class InputIt>
-   RVec(InputIt first, InputIt last) : fData(first, last) {}
+   RVec(Internal::VecOps::SmallVectorImpl<T> &&RHS) : Internal::VecOps::SmallVectorImpl<T>(N)
```
What other class might derive from `SmallVectorImpl<T>`? I get this in `llvm::SmallVector`'s case where the small buffer optimization size is a template parameter. But here?

eguiraud replied:
I removed this

Axel-Naumann commented on math/vecops/inc/ROOT/RVec.hxx:
```diff
@@ -276,163 +1214,179 @@ auto hpt = d.Define("pt", "sqrt(pxs * pxs + pys * pys)[E>200]")
    }
 
-   RVec<T> &operator=(std::initializer_list<T> ilist)
+   RVec &operator=(Internal::VecOps::SmallVectorImpl<T> &&RHS)
```
Here, too, I'd use `RVec` as argument type (unless I'm missing something)

eguiraud replied:
Solved by merging `RVec` and `SmallVectorImpl` (which has other advantages too)

Axel-Naumann commented on math/vecops/inc/ROOT/RVec.hxx:
```diff
@@ -100,27 +120,945 @@ auto MapFromTuple(Tuple_t &&t, std::index_sequence<Is...>)
+class SmallVectorBase {
+protected:
+   void *BeginX;
+   Size_T fSize = 0, Capacity;
```
To clearly see memory errors, consider
```suggestion
   Size_T fSize = 0;
   Size_T Capacity = -1;
```
To only make the declaration easier to read, consider
```suggestion
   Size_T fSize = 0;
   Size_T Capacity;
```

Axel-Naumann replied:
The coding convention differs here, not sure why. Maybe
```suggestion
   Size_T fSize = 0;
   Size_T fCapacity;
```

Axel-Naumann commented on math/vecops/inc/ROOT/RVec.hxx:
```diff
@@ -100,27 +120,945 @@ auto MapFromTuple(Tuple_t &&t, std::index_sequence<Is...>)
+protected:
+   void *BeginX;
+   Size_T fSize = 0, Capacity;
+   bool Owns = true;
```
Could this be `Capacity == -1`? I don't know what `Capacity == 12` and `fOwns == false` means - aren't these two mutually exclusive and this part of the same state description?

eguiraud replied:
we now use `fCapacity == 0` to indicate non-ownership. `clear()` calls `resetToSmall` in case of non-ownership, and `resetToSmall` now correctly sets `fCapacity` to the small vector capacity rather than to 0 as it happened in LLVM's SmallVector. The net effect is that `fCapacity == 0`  can now only be true if we are in memory adoption mode or if the small vector size is actually 0 (and it so happens that the behavior we want it these two cases is the same).

Axel-Naumann commented on math/vecops/inc/ROOT/RVec.hxx:
```diff
@@ -1,16 +1,33 @@
 #define ROOT_RVEC
 
+#if __cplusplus > 201402L
+#define RVEC_NODISCARD [[nodiscard]]
```
```suggestion
#define R__VEC_NODISCARD [[nodiscard]]
```
(ROOT's macros usually start with `R__`)

Axel-Naumann reviewed (commented):
Partial review, again - let's address `fOwns` first.

eguiraud commented on math/vecops/inc/ROOT/RVec.hxx:
```diff
@@ -1,16 +1,33 @@
  *************************************************************************/
 
+/*
+ * The implementation of RVec is based in LLVM's SmallVector.
```
this is not a doxygen comment, it's just a normal comment for developers opening the header

eguiraud commented:
I will make a push with the latest changes as soon as I solve the capacity-as-sentinel-for-ownership thing -- it's a bit tricky because the capacity is unsigned (so can't use -1 unless we change it to signed) and we can't use 0 because `clear()` happens to set the capacity and the size to 0 in the current implementation (and it's not easy to have `clear()` set the capacity to a more meaningful value without lowering a lot of the logic from `SmallVectorImpl` to `RVec`, as far as I can tell).

eguiraud commented on math/vecops/inc/ROOT/RVec.hxx:
```diff
@@ -100,27 +120,945 @@ auto MapFromTuple(Tuple_t &&t, std::index_sequence<Is...>)
+
+   reference front()
+   {
+      assert(!empty());
```
Changed to `throw` everywhere except in `operator[]` (hot function), where I changed them to `R__ASSERT`

eguiraud commented:
Rebased on master to roll back some of the changes and instead follow Axel's suggestions:

- `fSize` and `fCapacity` are now signed integers
- `fCapacity == -1` signals we are in "memory adoption" mode
- `assert`s changed into `throw`s where possible
- added a `ARCHITECTURE.md` document that summarizes the design and purpose of the VecOps library

I am having some issues with adding the small buffer size as a template parameter of `RVec`: templated helper functions would need to change signature and I/O needs patching.

Axel-Naumann commented:
> templated helper functions would need to change signature

yes, but I'd hope that this can be handled by passing `const baseclass &`?

> I/O needs patching

Indeed. I/O shouldn't care but doesn't know about that :-/

eguiraud commented:
> yes, but I'd hope that this can be handled by passing `const baseclass &`?

Yes, but we don't want to have to explain to users that they now have to program against `RVecImpl` in order to to use `RVec`.
Also some things become fairly ugly, e.g. all existing forward declarations of RVec will break, and even if you fix them to use two template parameters template functions that only see the forward declaration will not compile anymore: https://godbolt.org/z/vGYqvGe63 

I need to dedicate some brain time to decide if we are ok with this or not.

Axel-Naumann commented:
Good arguments, understood.

jblomer commented on math/vecops/ARCHITECTURE.md:
```diff
@@ -0,0 +1,103 @@
+- `operator[](mask)` has been added, as well as several other "numpy-like" helper
+  functions (these latter ones are free functions)
+- logical operators `==`, `<`, `>` etc. return vectors rather than booleans
+- the type of fSize and fCapacity is signed rather than unsigned, and fixed to 32 bits
```
@enrico Maybe that has been already discussed, but are we sure that we'll never support `RVec` > 2G?

eguiraud replied:
`fSize` is the number of elements, not the size in bytes of the object. we don't expect users to ever instantiate RVecs with size larger than 2^31 . what would the usecase be?

Axel-Naumann replied:
Can we detect the case and throw?

eguiraud replied:
We already do, see `grow` and `grow_pod`

enrico commented:
Guys, just a reminder to people when they tag: wrong Enrico !

eguiraud commented:
Rebased to fix conflicts, the last 3 commits are the only new ones. A type `RVecN` has been introduced which exposes the small buffer size as a template parameters, and `RVec` inherits publicly from it fixing the small buffer size to a reasonable value.

For now all mathematical operators and helper functions still deal with `RVec`, simply switching everything to taking _and returning_ `RVecN` instead is a breaking change.

pcanal commented on math/vecops/inc/ROOT/RVec.hxx:
```diff
@@ -1222,12 +1222,16 @@ public:
    {
+#ifndef NDEBUG
       R__ASSERT(idx < size());
+#endif
```
(only half kidding)
```suggestion
     assert(R__ASSERT(idx < size()));
```

eguiraud replied:
I like this in that it shows very clearly that `R__ASSERT` has a problem :smile: otoh I think readers will be puzzled when coming across this, I'd rather keep the more explicit ifndef

eguiraud commented:
Benchmark numbers in the PR description have been updated to reflect the latest changes. The PR is not a WIP anymore, the test failures in `pyroot_dependency_versions` and `roottest_python_cpp_cpp` are unrelated.

eguiraud commented:
@phsft-bot build

eguiraud commented:
Rebased to fix a minor conflict

Axel-Naumann commented:
The 32bit projectroot.math.vecops.test.gtest_math_vecops_test_vecops_rvec failure looks real?

eguiraud commented:
Totally, it seems to be due to my latest perf tweaks, I'm looking into it

eguiraud commented:
Uhm but it looks like it failed to fetch after my latest force-push? So I'm not 100% sure what it's building

eguiraud commented:
The testing condition was wrong, see last commit

bernhardmgruber commented on io/io/src/TGenCollectionProxy.cxx:
```diff
@@ -1043,13 +1043,14 @@ void* TGenCollectionProxy::At(UInt_t idx)
             fEnv->fIdx = idx;
             return &(fEnv->fLastValueVecBool);
          }
+      // intentional fall through
+      case ROOT::kROOTRVec:
```
Then mark it with `[[fallthrough]]`.

Axel-Naumann replied:
Unknown attribs are ignored but cause diagnostics that we don't want: https://www.godbolt.org/z/9Y3fKMb3n

bernhardmgruber commented on math/vecops/inc/ROOT/RVec.hxx:
```diff
@@ -11,6 +12,12 @@
 #ifndef ROOT_RVEC
 #define ROOT_RVEC
 
+#if __cplusplus > 201402L
+#define R__RVEC_NODISCARD [[nodiscard]]
+#else
+#define R__RVEC_NODISCARD
+#endif
+
```
Why not just use the attribute? Compilers are required to ignore unknown attributes.

eguiraud replied:
as per Axel's comment above

bernhardmgruber commented on math/vecops/inc/ROOT/RVec.hxx:
```diff
@@ -94,32 +111,985 @@ auto MapFromTuple(Tuple_t &&t, std::index_sequence<Is...>)
+   using Size_T = int32_t;
+
+protected:
+   void *fBeginX;
+   /// Always >= 0.
+   // Type is signed only for consistency with fCapacity.
+   Size_T fSize = 0;
+   /// Always >= -1. fCapacity == -1 indicates the RVec is in "memory adoption" mode.
+   Size_T fCapacity;
```
What is the reason for not initializing `fBeginX` and `fCapacity`, but `fSize`? It does not introduce overhead if you then initialize them differently in a ctor, but gives you safety in case you will add a ctor at some point.

eguiraud replied:
This comes from llvm's SmallVector implementation, so just guessing: on one hand there are no very good defaults here, it's _never_ ok  to have `fBeginX = nullptr`, for instance -- you always need the constructor to assign something sensible. And if you add a default that does not make sense it silences valgrind and static analyzers when you then forget to initialize to something sensible.

But now that you point it out I don't know what to do, how strongly do you feel about initializing these data members to values that most likely don't make sense? :smile:

bernhardmgruber replied:
I prefer determinism over the chance to catch an incorrect use of `SmallVector`'s API in valgrind. If `fBeginX` always needs to be initialized, then there should be no constructor that leaves it uninitialized and so giving it a member initializer incurs no cost but gives you potential safety in case someone adds a new ctor.

Axel-Naumann replied:
The LLVM folks argue exactly the inverse: according to them it's important to allow tooling to catch memory errors like uninitialized members. That's why [LLVM's code does not initialize these members](https://llvm.org/doxygen/SmallVector_8h_source.html#l00047).

bernhardmgruber replied:
Well, the LLVM folks also have `-ftrivial-auto-var-init=zero` which just let's the compiler do it. Do you turn that on for release builds of ROOT used in production? And if you leave the members uninitialized, will your users run the tools to catch their memory errors?

eguiraud replied:
again there is no value that makes sense there, everything is broken if those data members are not initialized by the constructor. as there are conflicting opinions on this I prefer sticking to the original implementation.

bernhardmgruber commented on math/vecops/inc/ROOT/RVec.hxx:
```diff
@@ -94,32 +111,985 @@ auto MapFromTuple(Tuple_t &&t, std::index_sequence<Is...>)
+#ifdef R__HAS_HARDWARE_INTERFERENCE_SIZE
+   constexpr std::size_t cacheLineSize = std::hardware_destructive_interference_size;
+#else
+   // safe bet: assume the typical 64 bytes
+   static constexpr std::size_t cacheLineSize = 64;
```
Is it really safe? The A64FX has 256bytes cachlines: https://www.fujitsu.com/downloads/SUPER/a64fx/a64fx_datasheet.pdf
But I guess you know your clients better and I do :)

eguiraud replied:
as safe as it gets, but we will eventually always use `std::hardware_destructive_interference_size`, users "just" need to get compilers that implement it

pcanal replied:
Using cmake configuration probing, could you start using it for the compiler that already have it? (99% guarantee that we would forget to make this update when C++17 becomes the ROOT default).

eguiraud replied:
I already use it for compilers that have it, see lines above these :D

bernhardmgruber commented on math/vecops/inc/ROOT/RVec.hxx:
```diff
@@ -94,32 +111,985 @@ auto MapFromTuple(Tuple_t &&t, std::index_sequence<Is...>)
+
+      // If this wasn't grown from the inline copy, deallocate the old space.
+      if (!this->isSmall())
+         free(this->begin());
```
```suggestion
         std::free(this->begin());
```
But that is certainly debatable. There are more occurences of this in this PR.

eguiraud replied:
this is how ROOT calls it :man_shrugging:

bernhardmgruber commented on math/vecops/inc/ROOT/RVec.hxx:
```diff
@@ -968,7 +2039,7 @@ template <typename T>
 auto Any(const RVec<T> &v) -> decltype(v[0] == true)
 {
    for (auto &&e : v)
-      if (e == true)
+      if (static_cast<bool>(e) == true)
```
Why not just `if (e)`?

eguiraud replied:
i have no idea, this code is from RVec 1.0. removed it, thanks

bernhardmgruber commented on math/vecops/test/vecops_rvec.cxx:
```diff
@@ -1191,3 +1088,95 @@ TEST(VecOps, Construct)
+
+// RVec does not guarantee exception safety, but we still want to test
+// that we don't segfault or otherwise crash if element construction or move throws.
+TEST(VecOps, NoExceptionSafety)
```
Could you add an additional test for a throwing copy constructor or a copy assignment operator? I believe this is more common than a throwing move ctor.

eguiraud replied:
done, let me know if you'd add something else

Axel-Naumann reviewed (approved):
Beautiful!
